### ISSUE TITLE
SparseCategoricalCrossentropy and Mixed Precision Training

### ISSUE THREAD

**Issue by zuyezheng** [tensorflow/tensorflow#50964]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): Yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Ubuntu 20.04
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device:
- TensorFlow installed from (source or binary):  Binary
- TensorFlow version (use command below): v2.5.0-0
- Python version: 3.8
- Bazel version (if compiling from source):
- GCC/Compiler version (if compiling from source):
- CUDA/cuDNN version: 11.2
- GPU model and memory: A6000

**Describe the current behavior**

`sparse_categorical_crossentropy` in `losses.py` performs an unnecessary [cast](https://github.com/tensorflow/tensorflow/blob/8c541bab667488249a34404170985dff8a9bab56/tensorflow/python/keras/losses.py#L1751) of `y_true` to `y_pred.dtype` since it's then [cast](https://github.com/keras-team/keras/blob/d3688b72924a4235598f0f80038de8c897f44799/keras/backend.py#L4952) to `int64` in `sparse_categorical_crossentropy` in `keras.backend.py`.

This seems to be the same code as in `categorical_crossentropy`, but causes issues with sparse, especially with mixed precision training and float16 as the loss in precision causes incorrect encodings or labels outside the domain resulting in incorrect or `nan` loss. With float16, issues start with a couple thousand labels and a couple hundred labels with bfloat16.

**Describe the expected behavior**

Cast to `y_pred.dtype` should be skipped.

**[Contributing](https://www.tensorflow.org/community/contribute)**

- Do you want to contribute a PR? (yes/no): 
- Briefly describe your candidate solution(if contributing):

**Standalone code to reproduce the issue**
Provide a reproducible test case that is the bare minimum necessary to generate
the problem. If possible, please share a link to Colab/Jupyter/any notebook.

https://colab.research.google.com/drive/1oRbNOnCo1i2HcXD2V4_-D1Bz2EVxiT65

**Comment by tilakrayal** [comment]
@zuyezheng ,

Please post this issue on [keras-team/keras](https://github.com/keras-team/keras/issues) repo.
To know more refer to:
[https://discuss.tensorflow.org/t/keras-project-moved-to-new-repository-in-https-github-com-keras-team-keras/1999](https://discuss.tensorflow.org/t/keras-project-moved-to-new-repository-in-https-github-com-keras-team-keras/1999)

**Comment by zuyezheng** [comment]
Created https://github.com/keras-team/keras/issues/15012, thanks.
